velox/vector/SelectivityVector.h

```diff
@@ -364,12 +364,46 @@ class SelectivityIterator {
 
 template <typename Callable>
 inline void SelectivityVector::applyToSelected(Callable func) const {
-  if (isAllSelected()) {
-    for (vector_size_t row = begin_; row < end_; ++row) {
-      func(row);
+  static const uint64_t kAllSet = -1;
+  if (allSelected_.has_value()) {
+    if (allSelected_.value()) {
+      for (vector_size_t row = begin_; row < end_; ++row) {
+        func(row);
+      }
+    } else {
+      bits::forEachSetBit(bits_.data(), begin_, end_, func);
     }
   } else {
-    bits::forEachSetBit(bits_.data(), begin_, end_, func);
+    bits::forEachWord(
```

Comment:
Is bits::forEachWord faster then bits::forEachSetBit ? I wonder why would that be the case given that forEachSetBit is implemented using forEachWord .

Reply:
I have added a brief explanation in the description above, let me know if that doesnt help.

Reply:
I wonder if the same improvement can be achieved by optimizing forEachBit, e.g. changing the following code to add fast path is word is all-set or all-unset.

```
     [isSet, bits, func](int32_t idx) {
        auto word = (isSet ? bits[idx] : ~bits[idx]);
        if (!word) {
          return;
        }
        while (word) {
          func(idx * 64 + __builtin_ctzll(word));
          word &= word - 1;
        }
      }
```

Reply:
To make sure we are on same page, I think you are suggesting we move this to logic to forEachBit. I think that will make forEachBit faster - though in isAllSelected, we will still pay cost of checking if all bits are selected. Would it make sense to add this same logic in both places ? Nevertheless let me move this logic there and see what performance change I observe.

Reply:
@mbasmanova Made the changes to forEachBit and I am seeing similar numbers:

```
Current implementation:
============================================================================
/Users/kpai/src/Velox/velox/vector/benchmarks/SelectivityVectorBenchmark.cpprelative  time/iter  iters/s
============================================================================
BM_isAllSelectedTrue(100000)                                24.02ms    41.64
BM_isAllSelectedFalseLastBit(100000)              49.76%    48.27ms    20.72
BM_isAllSelectedFalseFirstBit(100000)             49.89%    48.14ms    20.77
----------------------------------------------------------------------------
============================================================================

New implementation:
============================================================================
/Users/kpai/src/Velox/velox/vector/benchmarks/SelectivityVectorBenchmark.cpprelative  time/iter  iters/s
============================================================================
BM_isAllSelectedTrue(100000)                                26.20ms    38.16
BM_isAllSelectedFalseLastBit(100000)              96.88%    27.05ms    36.97
BM_isAllSelectedFalseFirstBit(100000)            102.44%    25.58ms    39.10
----------------------------------------------------------------------------
============================================================================
```